### PR TITLE
Added implementation for filemanager

### PR DESCRIPTION
**Pull request by williamlamjy**
added extra picture for seq between storage and decoder
added ref diagram for filemanager
added ref diagram for done command execution
link to dg here
https://williamlamjy.github.io/tp/DeveloperGuide.html

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -200,6 +198,44 @@ From the notes in the sequence diagram above, for every attribute in the instrum
...
+The loading of pre-existing data is mainly handled by the `filemanager` and `model` components. The main method calls `loadFileData(instrumentManager)` 
+in the `Storage` class which implements `InstrumentDecoder#readFile(instrumentManager, fileData)`. This method calls 
+`addSavedInstrumentToList(instrumentManager, textSegment)` for each pre-existing instrument which will add the corresponding 
+instrument in the `InstrumentManager` through calling the `addXYZToList` method in the XYZDecoder classes. In the event
+the instrument is not one of the 4 types of instruments, the `InstrumentDecoder` will throw a new `InvalidInstrumentInFileError`
+and display the corresponding error message.
+
+The figures below represents the sequence diagrams when the user loads a pre-existing crypto:
+
+<img src="images/FileManagerSeqBetweenStorageAndDecoder.png" width="800"/>
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:211]: ss1

I think you have to create a ref box in there to reference the bottom image. The bottom image is also formatted weirdly. Then the bottom image is to be encompassed in a sd ref frame

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -200,6 +198,44 @@ From the notes in the sequence diagram above, for every attribute in the instrum
...
+
+<img src="images/FileManagerSequenceDiagram.png" width="1040" height="800"/>
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:213]: Can use a darker orange for error box, it does not complement well with white font colour

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -200,6 +198,44 @@ From the notes in the sequence diagram above, for every attribute in the instrum
 prompt to get user to provide information for that attribute. This is done through a series of methods in
 the `TextUi` class.
 
+### Loading pre-existing data
+The loading of pre-existing data is mainly handled by the `filemanager` and `model` components. The main method calls `loadFileData(instrumentManager)` 
+in the `Storage` class which implements `InstrumentDecoder#readFile(instrumentManager, fileData)`. This method calls
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:203]:
```suggestion
in the `Storage` class which uses `InstrumentDecoder#readFile(instrumentManager, fileData)`. This method calls 
```
Can rephrase as u see fit, implements doesn't seem the most accurate here

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -200,6 +198,44 @@ From the notes in the sequence diagram above, for every attribute in the instrum
 prompt to get user to provide information for that attribute. This is done through a series of methods in
 the `TextUi` class.
 
+### Loading pre-existing data
+The loading of pre-existing data is mainly handled by the `filemanager` and `model` components. The main method calls `loadFileData(instrumentManager)` 
+in the `Storage` class which implements `InstrumentDecoder#readFile(instrumentManager, fileData)`. This method calls 
+`addSavedInstrumentToList(instrumentManager, textSegment)` for each pre-existing instrument which will add the corresponding
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:204]:
```suggestion
`InstrumentDecoder#addSavedInstrumentToList(instrumentManager, textSegment)` for each pre-existing instrument which will add the corresponding 
```
I think we should keep it consistent with this kind of annotation whenever we are talking about methods that are in a separate class that is not the main class (eg in this case the main class we talking about is Storage and InstrumentDecoder is the secondary class)

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -200,6 +198,44 @@ From the notes in the sequence diagram above, for every attribute in the instrum
...
+
+<img src="images/FileManagerSequenceDiagram.png" width="1040" height="800"/>
+
+More details about the reference frame for adding the decoded instrument into the `InstrumentManager` is shown below:
+
+<img src="images/FileManagerRefDiagCryptoDecoder.png" width="800"/>
+
+The process for loading other pre-existing instruments follow a similar process to the sequence above. The main difference
+would be the type of instrument decoder called, the different instrument specific decoded parameters and the type of instrument
+added to the `InstrumentManager`. For example when loading a stock instead of calling `addCryptoToList(textSegment, instrumentManager`
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:221]:
```suggestion
added to the `InstrumentManager`. For example when loading a stock instead of calling `CryptoDecoder#addCryptoToList(textSegment, instrumentManager`
```
Same here and for addStockToList as well

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -200,6 +198,44 @@ From the notes in the sequence diagram above, for every attribute in the instrum
...
+it will call `addStockToList(textSegement, instrumentManager)`.
+
+If loading the file data has any error, it will throw the corresponding file error. This file error will display the
+appropriate message through the `TextUi` class.
+
+### Storing current data
+The storing of current data is mainly handled by the `filemanager` and `model` components. The main method calls
+the `updateFileData(instruments)` in the Storage class which implements the `writeFile(instruments, writeToFile)` method
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:229]:
```suggestion
the `updateFileData(instruments)` in the Storage class which implements the `writeFile(instruments, writeToFile)` method
```

Similar issue here with the implements and InstrumentEncoder#writeFile

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -200,6 +198,44 @@ From the notes in the sequence diagram above, for every attribute in the instrum
...
+in the `InstrumentEncoder`. This method calls the `textFileFormatting()` method in the Instrument class for every instrument
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:230]:
```suggestion
in the `InstrumentEncoder`. This method calls the `Instrument#textFileFormatting()` method in the Instrument class for every instrument
```

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -200,6 +198,44 @@ From the notes in the sequence diagram above, for every attribute in the instrum
...
+
+The figures below represents the sequence diagrams when the user loads a pre-existing crypto:
+
+<img src="images/FileManagerSeqBetweenStorageAndDecoder.png" width="800"/>
+
+<img src="images/FileManagerSequenceDiagram.png" width="1040" height="800"/>
+
+More details about the reference frame for adding the decoded instrument into the `InstrumentManager` is shown below:
+
+<img src="images/FileManagerRefDiagCryptoDecoder.png" width="800"/>
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:217]: Regarding the frame can hold on first, I asking Rafael if the entities are suppose be inside the grid or can it be outside

**Review by theodorekwok — COMMENTED** [review]
Nice work, content is quite comprehensive. I think more implementation consideration can be added like talking about we have different decoder classes that helps to improve extensibility and to allow support for more instruments in the future. Or can also talk about improving cohesion through separating the decoder classes so that we can focus on each instrument type. I will need to update my parser section as well as I realised it can improved upon also

I was also reading the pr comments on our diagrams, they brought up a good point abt simplifying the seq diagram. So for your reference frame done command lets omit the self calls like getTypeIcon and getStatusIcon

File: docs/DeveloperGuide.md (hunk 8 above)
Comment (theodorekwok) [reply docs/DeveloperGuide.md:217]: Ok Rafael got back to me, the entities can be outside the grid, entities have to be inside the grid only if the obj is being created within the ref frame

**Comment by williamlamjy** [comment]
many self invocations and calls to validations in validate and decode methods not too sure if should omit those details and leave the diagrams as such or create an additional ref diagram to show the rs between validations and storage

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -227,6 +225,46 @@ From the notes in the sequence diagram above, for every attribute in the instrum
...
+the instrument is not one of the 4 types of instruments, the `InstrumentDecoder` will throw a new `InvalidInstrumentInFileError`
+and display the corresponding error message.
+
+The figures below represents the sequence diagrams when the user loads a pre-existing crypto:
+
+<img src="images/FileManagerSeqBetweenStorageAndDecoder.png" width="800"/>
+
+More details about the reference frame for decoding and updating the `InstrumentManager` is shown below:
+
+<img src="images/FileManagerSequenceDiagram.png" width="1080" height="500"/>
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:242]: InstrumentDecoder shld be <<class>>, same for CryptoDecoder as we are calling class level method addCryptoToList.
Also fix the constructor glitch in InvalidInstrumentInFileError

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -227,6 +225,46 @@ From the notes in the sequence diagram above, for every attribute in the instrum
...
+
+More details about the reference frame for adding the decoded instrument into the `InstrumentManager` is shown below:
+
+<img src="images/FileManagerRefDiagCryptoDecoder.png" width="950"/>
+
+The process for loading other pre-existing instruments follow a similar process to the sequence above. The main difference
+would be the type of instrument decoder called, the different instrument specific decoded parameters and the type of instrument
+added to the `InstrumentManager`. For example when loading a stock instead of calling `CryptoDecoder#addCryptoToList(textSegment, instrumentManager`
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:250]:
```suggestion
added to the `InstrumentManager`. For example when loading a stock instead of calling `CryptoDecoder#addCryptoToList(textSegment, instrumentManager)`
```

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -227,6 +225,46 @@ From the notes in the sequence diagram above, for every attribute in the instrum
...
+
+<img src="images/FileManagerSeqBetweenStorageAndDecoder.png" width="800"/>
+
+More details about the reference frame for decoding and updating the `InstrumentManager` is shown below:
+
+<img src="images/FileManagerSequenceDiagram.png" width="1080" height="500"/>
+
+More details about the reference frame for adding the decoded instrument into the `InstrumentManager` is shown below:
+
+<img src="images/FileManagerRefDiagCryptoDecoder.png" width="950"/>
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:246]: CryptoDecoder shld be <<class>>, the method calls are all class level

I think can remove the note, I don't quite get what you are saying. I think the diagram is comprehensive enough

File: docs/diagrams/FileManagerSeqBetweenStorageAndDecoder.puml [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+    end ref
+
+    InstrumentDecoder--> Storage
+    deactivate InstrumentDecoder
+    end
+
+[<--Storage
+deactivate Storage
+@enduml
```
Comment (theodorekwok) [inline docs/diagrams/FileManagerSeqBetweenStorageAndDecoder.puml:48]: Add newline

File: docs/diagrams/FileManagerRefDiagCryptoDecoder.puml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+activate InstrumentManager
+
+InstrumentManager--> CryptoDecoder
+deactivate InstrumentManager
+
+end
+
+deactivate CryptoDecoder
+
+@enduml
```
Comment (theodorekwok) [inline docs/diagrams/FileManagerRefDiagCryptoDecoder.puml:57]: Add newline

**Review by theodorekwok — COMMENTED** [review]
TLDR, need to check whether method calls are class level methods if they are the entity in seq diagram has a different label. Sry for catching this earlier, I also didn't know until yesterday

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -178,6 +178,10 @@ in the list of instruments:
 
 <img src="images/DoneCryptoSequenceDiagram.png" width="1040"/>
```
[the remark this answers is not in the document]
Comment (williamlamjy) [reply docs/DeveloperGuide.md:179]: ok the latest styles is pushed alrd right ?

[the remark this answers is not in the document]
Comment (theodorekwok) [reply docs/DeveloperGuide.md:179]: yep

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -178,6 +178,10 @@ in the list of instruments:
...
 
+More details about the reference frame for executing the done command is shown below:
+
+<img src="images/DoneCryptoExecuteDiagram.png" width="600"/>
+
 ### FileManager Component
 The `filemanager` package contains the `Storage`, `InstrumentEncoder` and `InstrumentDecoder` classes. It is defined in
```
Comment (theodorekwok) [inline docs/DeveloperGuide.md:186]: The XYZDecoder shld not have addStockToList but addXYZToList. And make the diagram bigger. Make the model diagram smaller.

**Review by theodorekwok — COMMENTED** [review]
Nice work! Just make the diagrams the right size. Generally we try to make the size of text in the diagram close to our normal text size in the dg

**Comment by theodorekwok** [comment]
LGTM!
